### ISSUE TITLE
Wrong cache size from ImageCache.default.calculateDiskStorageSize, larger than 10x

### ISSUE THREAD

**Issue by iaomw** [onevcat/Kingfisher#1825]
### Check List

Thanks for considering to open an issue. Before you submit your issue, please confirm these boxes are checked.

- [x] I have read the [wiki page](https://github.com/onevcat/Kingfisher/wiki) and [cheat sheet](https://github.com/onevcat/Kingfisher/wiki/Cheat-Sheet), but there is no information I need.
- [x] I have searched in [existing issues](https://github.com/onevcat/Kingfisher/issues?utf8=✓&q=is%3Aissue), but did not find a same one.
- [x] I want to report a problem instead of asking a question. It'd better to use [kingfisher tag in Stack Overflow](http://stackoverflow.com/questions/tagged/kingfisher) to ask a question.

### Issue Description
I noticed my cache size become much larger after using Kingfisher 7.0.0 version. Today, I did manually calculation. iPadOS 15.0 told me my app is taking 1.36GB documents & data. My app instance actually has nearly 1GB documents. But kingfisher is reporting nearly 5GB (5348560366) cache size on disk. That seems wrong.

#### Reproduce
The cache size looks also wrong on iOS 14.8, but it's not the same ratio. 

#### Other Comment
Not tested in iOS simulator

**Comment by onevcat** [comment]
Not sure if the documents & data number reported by iOS contains the cache data or not. Kingfisher is putting the disk cache into the Cache folder so maybe it is not counted by iOS? Can you confirm that the size of your Cache folder of your app?

Kingfisher is iterating all files under its disk cache folder by [this API](https://developer.apple.com/documentation/foundation/url/1780058-resourcevalues), with the [`fileSizeKey` passed in](https://developer.apple.com/documentation/foundation/urlresourcekey/1417747-filesizekey). 

I guess there would quite little we can do if this API is returning a wrong value for a file. But I will try to confirm it first.

**Comment by iaomw** [comment]
Thanks, I did a quick test with this API and Kingfisher source. This API seems not working for folder `com.onevcat.Kingfisher.ImageCache.default` . So I cannot compare the folder size with total size of all individual files.

While the folder is on iPad, it's difficult to get size directly. I will test with iOS simulator later.

I reported this issue, because Kingfisher 6.x works fine to get total size of cache. I didn't count it manually, but it was very close roughly. Any related code changed from 6.x to 7.0.

**Comment by onevcat** [comment]
> because Kingfisher 6.x works fine to get total size of cache

Oh, that is surprising! I believe there is no change around this. So I guess it is more likely an iOS change... :(

Any chance to try it on an iOS 14 device or so?

**Comment by iaomw** [comment]
Confirmed on macOS and iOS simulator. The folder size on macOS and simulator are exactly the same.

If the API works correctly, I am confused what's taking 5GB space in the cache folder on iPad. 
Even if I pre-cached all the image from the server domain, it's only a bit larger 800MB.

**Comment by iaomw** [comment]
The issue has been located. However, it still behaves differently under different case. 

For code simplicity, I am using Kingfisher for both online and local image with URL.

Kingfisher will copy the local image to cache folder, that's strange but not a major issue. 
- If the app is killed, then Kingfisher will not copy the local image again while displaying the image. (not 100% sure)
- If the app is re-launched from Xcode, Kingfisher will copy the local image again into cache folder.

As the result, Kingfisher cache folder will grow rapidly with local images.  For some reason, launch from Xcode break the duplication check for Kingfisher local images. 


You may find some huge sized test image from the link below. 4K & 8K.
https://ambientcg.com/api/v1/full_json?type=hdri

**Comment by iaomw** [comment]
The recommended API `LocalFileImageDataProvider` also has this issue.

**Comment by onevcat** [comment]
Thanks for the investigation.

It is the way that Kingfisher would also copy a local file to the cache folder (although this is a thing to optimize). The cache manager is not distinguishing the source of image, but just cache them.

The cache works based on the cache key. So if you have the same cache key, that should not happen. May I know how did you load the local files? Can you share some code snippet?

**Comment by iaomw** [comment]
```
ForEach(0..<urls.count) { idx in
                        
                        let url = urls[idx]
                        let provider = LocalFileImageDataProvider(fileURL: url)
                        
                        HStack {
                            VStack(alignment: .leading) {
                                   AnotherView() 
                            }
                            
                            Spacer()
                            
                            KFImage.dataProvider(provider)
                            //KFImage.url(url)
                                .scaleFactor(UIScreen.main.scale)
                                .backgroundDecode()
                                .fade(duration: 0.2)
                                .cancelOnDisappear(true)
                                .placeholder {ProgressView()}
                                //.downsampling(size: <#T##CGSize#>)
                                //.cacheMemoryOnly() // Could solve the file copy problem, not sure about memory usage loading 8K image
                                .resizable().scaledToFit()
                            
                        }.frame(height: 100)
```

**Comment by onevcat** [comment]
> If the app is killed, then Kingfisher will not copy the local image again while displaying the image. (not 100% sure)
If the app is re-launched from Xcode, Kingfisher will copy the local image again into cache folder.
> As the result, Kingfisher cache folder will grow rapidly with local images. For some reason, launch from Xcode break the duplication check for Kingfisher local images.

I can confirm this behavior in a simulator when running with Xcode 13. Just did some investigation, it seems that every time you run the app, Xcode creates a new app folder under the `~/Library/Developer/CoreSimulator/Devices/${device_id}/data/Containers/Bundle/Application` folder (if you open that folder and inspect it while stopping old app and running again, you will find the app folder is replaced). This causes your local images being cached again and again while starting running your apps, because the file URL (which is also the `cacheKey` of a resource by default) is changing every time.

The similar behavior is also confirmed on a real device. The means when an upgrade version of this app is installed on the device, the old caches for local file might just be invalidated. The good news is this issue won't happen until an upgrade was made, and the disk cache has a 7 days expiration duration by default. So it won't cause a huge problem.  But actually it is not an ideal solution to use the full URL of a local file to determine cache status anymore.

A simple way now for framework users to solve it now, is setting your own `cacheKey` for the local resource:

```swift
let provider = LocalFileImageDataProvider(fileURL: fileURL, cacheKey: "my_cache_key_for_fileURL")
UIImageView().kf.setImage(with: provider)
```

It should be possible to drop the changeable part and use only the file name and query part as the cache key. I will consider if we can have this by default and if this would work better than current model.

**Comment by damirstuhec** [comment]
@onevcat can you please make `localFileCacheKey` a public API? There's currently no way to clear cache for a local file image since `localFileCacheKey` is defined internally. I would like to do `ImageCache.default.removeImage(forKey: imageURL.localFileCacheKey)`. Thanks in advance!

**Comment by onevcat** [comment]
@damirstuhec 
Maybe it is possible to do it by `LocalFileImageDataProvider(fileURL: imageURL).cacheKey`. The `localFileCacheKey` is an extension in `URL` now and exposing it is a bit more invasive, so we'd better act with caution.

**Comment by onevcat** [comment]
@damirstuhec Now, the `cacheKey` of a local URL should return the correct default cache key used in local cache with the fix in #1914. So you can just use `ImageCache.default.removeImage(forKey: imageURL.cacheKey)` for both remote and local URLs. I will tag it as a patch soon.

**Comment by damirstuhec** [comment]
Awesome. Thanks!
